Apromore-Calendar/src/main/java/org/apromore/calendar/model/CalendarModel.java

```diff
@@ -169,67 +126,29 @@ public DurationModel getDuration(Long starDateTimeUnixTs, Long endDateTimeunixTs
     if (starDateTimeUnixTs == null || starDateTimeUnixTs.length == 0 || endDateTimeunixTs == null
         || endDateTimeunixTs.length == 0)
       return new long[] {};
-
     long[] resultList = new long[starDateTimeUnixTs.length];
-
-    ZoneId zone = ZoneId.of(zoneId);
-
     IntStream.range(0, starDateTimeUnixTs.length).parallel().forEach(i -> {
-
-      ZonedDateTime zonedStartDateTime =
-          ZonedDateTime.ofInstant(Instant.ofEpochMilli(starDateTimeUnixTs[i]), zone);
-      ZonedDateTime zonedEndDateTime =
-          ZonedDateTime.ofInstant(Instant.ofEpochMilli(endDateTimeunixTs[i]), zone);
-      resultList[i] = getDuration(zonedStartDateTime, zonedEndDateTime).getDuration().toMillis();
-
+      resultList[i] = getDurationMillis(Instant.ofEpochMilli(starDateTimeUnixTs[i]),
+                                  Instant.ofEpochMilli(endDateTimeunixTs[i]));
     });
 
     return resultList;
   }
 
-
-
-  public void populateHolidayMap() {
-    if (holidayLocalDateMap.isEmpty()) {
-      holidayLocalDateMap = holidays.parallelStream().collect(
-          Collectors.toMap(HolidayModel::getHolidayDate, Function.identity(), (e1, e2) -> e1));
-    }
+  private List<ChronoInterval<Moment>> getWorkDayIntervals(Instant start, Instant end) {
+    return workDays.stream()
+            .filter(WorkDayModel::isWorkingDay)
+            .map(workDay -> workDay.getRealIntervals(start, end, ZoneId.of(zoneId)).stream())
+            .flatMap(Function.identity())
+            .collect(Collectors.toList());
   }
 
-
-
-  private void populateWorkDayMap() {
-    if (dayOfWeekWorkDayMap.isEmpty()) {
-      dayOfWeekWorkDayMap = workDays.parallelStream()
-          .collect(Collectors.toMap(WorkDayModel::getDayOfWeek, Function.identity()));
-    }
-  }
-
-
-  private boolean isEndDay(ZonedDateTime endDateTime, ZonedDateTime dayDateTime) {
-    return dayDateTime.toLocalDate().isEqual(endDateTime.toLocalDate());
-  }
-
-
-  private boolean isStartDay(ZonedDateTime zonedStartDateTime, ZonedDateTime dayDateTime) {
-    return isEndDay(zonedStartDateTime, dayDateTime);
-  }
-
-
-  private boolean isSameDay(LocalDate localStartDate, LocalDate localEndDate) {
-    return localStartDate.equals(localEndDate);
-  }
-
-
-  private Duration getDurationForSameDay(OffsetTime startTime, OffsetTime endTime,
-      WorkDayModel workDayModel) {
-    DurationModel durationModel = new DurationModel();
-
-    Duration duration = Duration.between(workDayModel.getAdjustedStartTime(startTime),
-        workDayModel.getAdjustedEndTime(endTime));
-
-    return duration.isNegative() ? Duration.ZERO : duration;
-
+  private List<ChronoInterval<Moment>> getHolidayIntervals() {
+    return holidayIntervals != null
+            ? holidayIntervals
+            : (holidayIntervals = holidays.stream()
```

Comment:
This evaluation mixed with assignment of `holidayIntervals` is tricky enough that I'd request it be changed.

Reply:
Done